torch/serialization.py

```diff
@@ -406,28 +407,75 @@ def persistent_id(obj: Any) -> Optional[Tuple]:
                               "for correctness upon loading.")
             return ('module', obj, source_file, source)
 
-        elif torch.is_storage(obj):
+        if isinstance(obj, torch.storage.TypedStorage) or torch.is_storage(obj):
+            if isinstance(obj, torch.storage.TypedStorage):
+                # TODO: Once we decide to break serialization FC, this case
+                # can be deleted
+                storage = obj.storage
+                storage_type_str = obj.pickle_storage_type()
+                storage_type = getattr(torch, storage_type_str)
+                dtype = obj.dtype
+                storage_numel = obj.numel()
+
+            else:
+                storage = obj
+                storage_type = normalize_storage_type(type(obj))
+                dtype = torch.uint8
+                storage_numel = cast(Storage, storage).nbytes()
+
             view_metadata: Optional[Tuple[str, int, int]]
-            obj = cast(Storage, obj)
-            storage_type = normalize_storage_type(type(obj))
+            storage = cast(Storage, storage)
```

Comment:
instead of casting here, it seems better to `assert isinstance(obj, Storage)` in else branch and then mypy should be able to figure it out